---
title: "301-2017-09-18-taking-apart-example_fn"
author: ''
output:
  html_document:
    css: ~/Documents/headings.css
    fig_caption: yes
    highlight: default
    number_sections: yes
    theme: readable
    toc: yes
  pdf_document:
    highlight: tango
    number_sections: yes
    pandoc_args: --variable=geometry:margin=0.75in
    toc: no
fontsize: 10pt
---
<A NAME="top"> </A>

```{r global_options, include=FALSE}
require(knitr)
opts_chunk$set(eval = TRUE, echo = TRUE, fig.height = 8, fig.width = 8, tidy=FALSE)
```

`r getwd()`  

Last knit was `r Sys.time()`

Copied from "209-2017-09-12-example-containing-exclusion_region.Rmd"


# A little history

The two data sets in `gatars`:

`alternative_example` and `null_example` I will be creating here.  

There is some history.  From `lacie/aaa/packages/gritsr2/gails-stuff/a-chronicle/04-v-0.1.4.html` I see that gritsr2 contains the following four data sets
```
bim_original.RData
exclusion_region.RData
genotype_bim_original.RData
hotspot.RData
```
which I moved from `d-2016-04-20/06-adding-valeries-data/gritsr_packages_RData`

These files were created by “107-2016-10-24-description-adding-valeries-data.Rmd”

The code below, I copied from "lacie/aaa/packages/gritsr2/gails-stuff/b-examples/03-v-0.1.4/a-build-hg38/201-gritsr-example-dot-RData"

I need to make sure I have an exclusion region

I need to make sure I use either `start` or `begin` but not both.


```{r}
library(gritsr2)
set.seed(1)
genotype_bim_original = genotype_bim_original[1:5, ]
params_example = params_example_fn(
  bim_original, genotype_bim_original, hotspot,
  NNN = 200,
  p_subset_genotype = .2,
  beta_causal_snps = .85,
  causal_snps = 3:5)
NNN = 200
p_subset_genotype = .2
beta_causal_snps = .85
causal_snps = 3:5
```


# `params_example_fn`

`p_sampling_set_generator` is a vector of length `nrow(bim_original = 124303`  uniform random numbers between 0.002 and 0.1

I am going to generate `NNN/4` affected sib pairs and `NNN/2` unaffected controls


```{r}
N_hap_rows = 5000
age_generator = data.frame(mean = c(65, 70), sd = 8)
row.names(age_generator) = c("FAMILY", "CASE_CONTROL")
str(bim_original)

dimnames_dosage_2 = bim_original$snp
str(dimnames_dosage_2)
dimnames_genotype_dosage_2 = genotype_bim_original$snp
str(dimnames_genotype_dosage_2)
beta_nongenetic = c(1.2, -0.02, -3.1, 0.07)
beta_causal_snps = rep(beta_causal_snps, length(causal_snps))
N_affected_sib_pairs = NNN/4
N_independent_cases = 0
N_independent_controls = NNN/2
N_unaffected_sib_pairs = 0
p_sampling_set_generator = round(runif(nrow(bim_original), min = 0.002, max = 0.1), 3)
params_sim = list(
  bim_original = bim_original,
  genotype_bim_original = genotype_bim_original,
  NNN = NNN,
  N_hap_rows = N_hap_rows,
  N_smaller = length(dimnames_dosage_2),
  p_sampling_set_generator = p_sampling_set_generator,
  p_subset_genotype = p_subset_genotype,
  
  age_generator = age_generator,
  beta_nongenetic = beta_nongenetic,
  beta_causal_snps = beta_causal_snps,
  causal_snps = causal_snps,
  N_affected_sib_pairs = N_affected_sib_pairs,
  N_independent_cases = N_independent_cases,
  N_independent_controls = N_independent_controls,
  N_unaffected_sib_pairs = N_unaffected_sib_pairs)
```


# `example_genotype_etc_fn`

```{r}
N_hap_rows = params_example$N_hap_rows
p_sampling_set_generator = params_example$p_sampling_set_generator
NNN = params_example$NNN
p_subset_genotype = params_example$p_subset_genotype
bim_original = params_example$bim_original
str(bim_original)
genotype_bim_original = params_example$genotype_bim_original
str(genotype_bim_original)
dimnames_genotype_2 = bim_original$snp
target_markers = genotype_bim_original$snp

haplotypes = do.call(cbind, lapply(p_sampling_set_generator, function(ppp){
  rbinom(N_hap_rows, 1, ppp)
}))
str(haplotypes)


null_model_chromosomes = null_model_chromosomes_fn(params_example, 1)[[1]]
str(null_model_chromosomes)
genotype_0 = do.call(rbind, lapply(1:nrow(null_model_chromosomes), function(kkk){
  this_row = null_model_chromosomes[kkk, ]
  haplotypes[this_row[1], ] + haplotypes[this_row[2], ]
}))
dimnames(genotype_0) = list(1:NNN, dimnames_genotype_2)
genotype_1 = genotype_0[, colSums(genotype_0) > 0]
choose_these = dimnames(genotype_1)[[2]] %in% target_markers |
  as.logical(rbinom(1:ncol(genotype_1), 1, p_subset_genotype))
genotype = genotype_1[, choose_these]
bim = bim_original[bim_original$snp %in% dimnames(genotype)[[2]], ]
# genotype_dosage = dosage[, dimnames(dosage)[[2]] %in%  dimnames_genotype_dosage_2]
# genotype_bim = genotype_bim_original[genotype_bim_original$snp %in% dimnames(genotype_dosage)[[2]], ]
list(bim = bim,
     genotype = genotype,
     target_markers = target_markers)
}

# I am using the new terminology dosage (old) = genotype (new)
# genotype_bim and genotype_dosage (old) replaced with target_markers
# example_fn will need to be adjusted.

```

# `null_model_chromosomes_fn = function(params, N_chromosomes){`
This is from 2016-12-22

```{r}
N_chromosomes = 1
params = params_example
  N_affected_sib_pairs = params$N_affected_sib_pairs
  N_independent_cases = params$N_independent_cases
  N_independent_controls = params$N_independent_controls
  N_unaffected_sib_pairs = params$N_unaffected_sib_pairs
  lapply(1:N_chromosomes, function(n_chromosome){
    affected_sib_pairs = if(N_affected_sib_pairs == 0){
      NULL
    } else {
      sib_pairs_zero_fn(params, N_affected_sib_pairs)
    }
    independent_cases = if(N_independent_cases == 0){
      NULL
    } else {
      independent_people_zero_fn(params, N_independent_cases)
    }
    independent_controls = if(N_independent_controls == 0){
      NULL
    } else {
      independent_people_zero_fn(params, N_independent_controls)
    }
    unaffected_sib_pairs = if(N_unaffected_sib_pairs == 0){
      NULL
    } else {
      sib_pairs_zero_fn(params, N_unaffected_sib_pairs)
    }
    answer_0 = rbind(
      affected_sib_pairs, 
      independent_cases, 
      unaffected_sib_pairs,
      independent_controls)
    rownames(answer_0) = NULL
    answer = answer_0
    answer
  })
} 
```


# `null_model_chromosomes_fn`

```{r}
null_model_chromosomes_fn = function(params_generating, N_chromosomes){
  lapply(1:N_chromosomes, function(n_chromosome){
    one_null_model_chromosome_fn(params_generating, y_needed = FALSE)
  })
} 
```

# `one_null_model_chromosome_fn`


```{r}
params_generating = params_example
y_needed = FALSE
N_affected_sib_pairs = params_generating$N_affected_sib_pairs
N_independent_cases = params_generating$N_independent_cases
N_independent_controls = params_generating$N_independent_controls
N_unaffected_sib_pairs = params_generating$N_unaffected_sib_pairs
affected_sib_pairs = if(N_affected_sib_pairs == 0){
  NULL
} else {
  sib_pairs_zero_fn(params_generating, N_affected_sib_pairs)
}
independent_cases = if(N_independent_cases == 0){
  NULL
} else {
  independent_people_zero_fn(params_generating, N_independent_cases)
}
independent_controls = if(N_independent_controls == 0){
  NULL
} else {
  independent_people_zero_fn(params_generating, N_independent_controls)
}
unaffected_sib_pairs = if(N_unaffected_sib_pairs == 0){
  NULL
} else {
  sib_pairs_zero_fn(params_generating, N_unaffected_sib_pairs)
}
haplotype_indices = rbind(
  affected_sib_pairs, 
  independent_cases, 
  unaffected_sib_pairs,
  independent_controls)
rownames(haplotype_indices) = NULL

if(!y_needed){
  list(haplotype_indices = haplotype_indices)
} else {
  affected_sib_pairs = if(N_affected_sib_pairs == 0){
    NULL
  } else {
    y_cases_fn(params_generating, N_affected_sib_pairs * 2)
  }
  independent_cases = if(N_independent_cases == 0){
    NULL
  } else {
    y_cases_fn(params_generating, N_independent_cases)
  }
  independent_controls = if(N_independent_controls == 0){
    NULL
  } else {
    y_controls_fn(params_generating, N_independent_controls)
  }
  unaffected_sib_pairs = if(N_unaffected_sib_pairs == 0){
    NULL
  } else {
    y_controls_fn(params_generating, N_unaffected_sib_pairs * 2)
  }
  y = c(affected_sib_pairs, independent_cases, unaffected_sib_pairs, independent_controls)
  list(haplotype_indices = haplotype_indices,
       y = y)
}
} 

```


# `sib_pairs_zero_fn`

Returns a list containing 2 objects.

`family_id` are the family ids 

`haplotype_indices is a matrix with `N_sib_pairs_each * 2` rows and 2 columns named `h1` and `h2`.
The `n`-th row contains the two haplotype indices assigned to the `n`-th person.
Haplotype indices can range between `1` and `N_hap_rows`.

```{r}
N_sib_pairs_each = N_affected_sib_pairs
N_hap_rows = params_generating$N_hap_rows
many_person_haplotypes = sample(
  1:N_hap_rows, 
  size = 4 * N_sib_pairs_each, 
  replace = TRUE)
many_two_parent_haplotypes = split_fn(
  many_person_haplotypes, 4)
N_sibs_in_each_family = 2
sib_pairs_zero_0 = lapply(many_two_parent_haplotypes, function(this_foursome){
  mother = this_foursome[1:2]
  father = this_foursome[3:4]
  lapply(1:N_sibs_in_each_family, function(dummy){
    haplotype_indices = c(
      mother[sample(1:2, 1)], 
      father[sample(1:2, 1)])
    haplotype_indices 
  })
})
family_id = unlist(lapply(1:length(sib_pairs_zero_0), function(kkk){
  this_family = sib_pairs_zero_0[[kkk]]
  rep(kkk, length(this_family))
}))
sib_pairs_zero_almost = do.call(c, sib_pairs_zero_0)
sib_pairs_zero = do.call(rbind, sib_pairs_zero_almost) 
dimnames(sib_pairs_zero)[[2]] = c("h1", "h2")
sib_pairs_zero_fn_answer = list(family_id = family_id,
     haplotype_indices = sib_pairs_zero)
str(sib_pairs_zero_fn_answer)
```




# `example_fn = function(params_example){`



```{r}

eee = example_genotype_etc_fn(params_example)
bim = eee$bim
genotype = eee$genotype
target_markers = intersect(eee$target_markers, dimnames(genotype)[[2]])
g_target = genotype[, target_markers]
Psi = Psi_fn(params_example, Psi_override = FALSE)
fam = example_fam_fn(params_example, g_target)
example_genotype_etc_fn_answer = list(bim = bim,
     genotype = genotype,
     exclusion_region = exclusion_region,
     fam = fam,
     target_markers = target_markers,
     Psi = Psi)
```


# `example_genotype_etc_fn = function(params_example){`

```{r}
N_hap_rows = params_example$N_hap_rows
p_sampling_set_generator = params_example$p_sampling_set_generator
NNN = params_example$NNN
p_subset_genotype = params_example$p_subset_genotype
bim_original = params_example$bim_original
genotype_bim_original = params_example$genotype_bim_original
dimnames_genotype_2 = bim_original$snp
target_markers = genotype_bim_original$snp

haplotypes = do.call(cbind, lapply(p_sampling_set_generator, function(ppp){
  rbinom(N_hap_rows, 1, ppp)
}))
null_model_chromosomes = null_model_chromosomes_fn(params_example, 1)[[1]]
str(null_model_chromosomes)




null_model_chromosomes
```



```{r}
genotype_0 = do.call(rbind, lapply(1:nrow(null_model_chromosomes), function(kkk){
  this_row = null_model_chromosomes[kkk, ]
  haplotypes[this_row[1], ] + haplotypes[this_row[2], ]
}))
dimnames(genotype_0) = list(1:NNN, dimnames_genotype_2)
genotype_1 = genotype_0[, colSums(genotype_0) > 0]
choose_these = dimnames(genotype_1)[[2]] %in% target_markers |
  as.logical(rbinom(1:ncol(genotype_1), 1, p_subset_genotype))
genotype = genotype_1[, choose_these]
bim = bim_original[bim_original$snp %in% dimnames(genotype)[[2]], ]
# genotype_dosage = dosage[, dimnames(dosage)[[2]] %in%  dimnames_genotype_dosage_2]
# genotype_bim = genotype_bim_original[genotype_bim_original$snp %in% dimnames(genotype_dosage)[[2]], ]
example_genotype_etc_fn_answer = list(bim = bim,
                                      genotype = genotype,
                                      target_markers = target_markers)


# I am using the new terminology dosage (old) = genotype (new)
# genotype_bim and genotype_dosage (old) replaced with target_markers
# example_fn will need to be adjusted.

```




```{r}
bim = alternative_example$bim
genotype = alternative_example$genotype
fam = alternative_example$fam
target_markers = alternative_example$target_markers
Psi = alternative_example$Psi
```

```{r echo = TRUE}
str(bim)
str(genotype)
str(fam)
str(target_markers)
str(Psi)
```

# `null_example`

```{r}
set.seed(1)
genotype_bim_original = genotype_bim_original[1:5, ]
params_example = params_example_fn(
  bim_original, genotype_bim_original, hotspot,
  NNN = 200,
  p_subset_genotype = .2,
  beta_causal_snps = 0,
  causal_snps = 3:5)
null_example = example_fn(params_example)
null_example$params_example = params_example
save(null_example, file = "null_example.RData")
```


```{r}
bim = null_example$bim
genotype = null_example$genotype
fam = null_example$fam
target_markers = null_example$target_markers
Psi = null_example$Psi
```

```{r echo = TRUE}
str(bim)
str(genotype)
str(fam)
str(target_markers)
str(Psi)
```
