---
title: "305-2017-11-02-example-after-basic_and_optimized_lu_fn"
author: ''
output:
  html_document:
    css: ~/Documents/headings.css
    fig_caption: yes
    highlight: default
    number_sections: yes
    theme: readable
    toc: yes
  pdf_document:
    highlight: tango
    number_sections: yes
    pandoc_args: --variable=geometry:margin=0.75in
    toc: no
fontsize: 10pt
---
<A NAME="top"> </A>

```{r global_options, include=FALSE}
require(knitr)
opts_chunk$set(eval = TRUE, echo = TRUE, fig.height = 8, fig.width = 8, tidy=FALSE)
```

`r getwd()`  

Last knit was `r Sys.time()`


# A little history for the 305 file

The data set in `gatars`: `gatars_example`  I will be creating here.  

There is some history.  From `lacie/aaa/packages/gritsr2/gails-stuff/a-chronicle/04-v-0.1.4.html` I see that gritsr2 contains the following four data sets
```
bim_original.RData
exclusion_region.RData
genotype_bim_original.RData
hotspot.RData
```

`bim_original` is a data.frame containing the columns `chromosome`, `snp`, and `bp` and 124303 rows, so these are possibilities for sampling sets.


`genotype_bim_original` is a data.frame containing the columns `chromosome`, `snp`, and `bp` and 47 rows, so these are possibilities for target markers.


`exclusion_region` is a data.frame containing the columns `chromosome` `begin` and `end` and 113 rows.

`hotspot` is a data.frame with 25657 rows and the columns `chromosome`, `center`, `start` and `end`.



which I moved from `d-2016-04-20/06-adding-valeries-data/gritsr_packages_RData`

These files were created by “107-2016-10-24-description-adding-valeries-data.Rmd”

The code below, I copied from "lacie/aaa/packages/gritsr2/gails-stuff/b-examples/03-v-0.1.4/a-build-hg38/201-gritsr-example-dot-RData"


```{r}
library(gatars)
library(gritsr2)
set.seed(31)
RData_file = "gatars_example.RData"
source_0 = getwd()
source_1 = strsplit(source_0, "/")[[1]]
source = paste0(paste(source_1[(length(source_1) - 2):length(source_1)], collapse = "/"),
      "/305-2017-11-02-example-after-basic_and_optimized_lu_fn.Rmd")
genotype_bim_original = genotype_bim_original[1:5, ]
params_example = params_example_fn(
  bim_original, genotype_bim_original, hotspot,
  NNN = 200,
  p_subset_genotype = .2,
  beta_causal_snps = .85,
  causal_snps = 3:5)
gatars_example = example_fn(params_example)
names(gatars_example)
str(gatars_example$bim)
str(gatars_example$genotype)
str(gatars_example$exclusion_region)
str(gatars_example$fam)
gatars_example$target_markers
str(gatars_example$Psi)
gatars_example$Psi[1:10, 1:10]
gatars_example$params_example = params_example

additional_attributes = list(
  RData_file = RData_file,
  when_created = Sys.time(),
  source = source
)
attributes(gatars_example) = c(attributes(gatars_example),
                                    additional_attributes)
                                               
save(gatars_example, file = RData_file)
```

```{r}
# here i am calling `gatars`
bim = gatars_example$bim
exclusion_region = gatars_example$exclusion_region
fam = gatars_example$fam
genotype = gatars_example$genotype
target_markers = gatars_example$target_markers[3:5]
Psi = gatars_example$Psi
str(Psi)
NNN = nrow(fam)
first_ten = 1:10
last_ten = NNN - (9:0)
matrix_image_fn(Psi[c(first_ten, last_ten), c(first_ten, last_ten)],
                main = "First and last 10 rows and columns of Psi")

# Checking the rank of the genotype_target_markers matrix
library(Matrix)
genotype_target_markers = genotype[, target_markers]
list(target_markers = target_markers,
     rank = as.numeric(rankMatrix( genotype_target_markers)))

# Creating the sampling_set
set.seed(1)
epsilon_on_log_scale = 0.01
exclusion_region = NULL
sampling_set = gatars_sampling_set(
  bim,
  epsilon_on_log_scale,
  exclusion_region,
  genotype,
  hotspot,
  target_markers
)
print(sampling_set)


```



Inside `gatars`

```{r}
N_sim_reps = 100
weights = NULL
params_sampling_set = sampling_set$params_sampling_set
sampling_set = sampling_set$sampling_set
alpha_uni_N_increments = 10
params = params_fn(
  alpha_uni_N_increments, params_sampling_set, fam, Psi, sampling_set, N_sim_reps, weights)
  # ooo = uno_experimento_fn(params, calculate_optimized = TRUE)
```

Inside `uno_experimento_fn`


```{r}
#' @export
# uno_experimento_fn = function(params, calculate_optimized){
  adaptive_conf_level = params$adaptive_conf_level
  alpha_uni = params$alpha_uni
  g_target = params$g_target
  MMM = params$MMM
  N_sim_reps_interval = params$N_sim_reps_interval
  N_sim_reps_limit = params$N_sim_reps_limit
  theta_init = params$theta_init
  WWW = params$WWW
  y_1 = params$yyy
  y_2 = params$e_y
  Phi = params$Phi
  sampling_set = params$sampling_set
  test_size = params$test_size
  theta_init = params$theta_init
```


Inside `basic_and_optimized_lu_fn`

```{r}
#' @export
basic_and_optimized_lu_fn = function(alpha_uni, g_target, Phi, theta, WWW, y_1, y_2){
  # ---------------------------------------------------- zzz etc
  zzz_etc = zzz_etc_fn(g_target, Phi, WWW, y_1, y_2)
  zzz = zzz_etc$zzz
  mu_z = zzz_etc$mu_z
  V_z = zzz_etc$V_z
```

Inside `zzz_etc_fn`

```{r}
  z_1 = as.vector(WWW %*% t(g_target) %*% y_1)
  z_2 = as.vector(WWW %*% t(g_target) %*% y_2)
  zzz = c(z_1, z_2)
  
  e_g_target_1 = colMeans(g_target)
  e_g_target = matrix(rep(e_g_target_1, nrow(g_target)), nrow = nrow(g_target), byrow = TRUE)
  e_z_1 = matrix(as.vector(WWW %*% t(e_g_target) %*% y_1), ncol = 1)
  e_z_2 = matrix(as.vector(WWW %*% t(e_g_target) %*% y_2), ncol = 1)
  mu_z = rbind(e_z_1, e_z_2)
  
  V_G = cov(g_target)
  W_VG_W = WWW %*% V_G %*% WWW
  V_z = kronecker(Phi, W_VG_W)
  
  list(zzz = zzz,
       mu_z = mu_z,
       V_z = V_z)
```



  # -------------------------------------------------- basic_AAA
  basic_ones = alpha_uni[apply(alpha_uni, 1, function(row){
    sum(row == 0) > 1
  }), ]
  MMM = ncol(g_target)
  AAA_basic = lapply(1:nrow(basic_ones), function(jjj){
    alpha = basic_ones[jjj, ]
    AAA = AAA_fn(alpha$B, alpha$S, alpha$T, MMM)
    list(alpha = alpha, AAA = AAA)
  })
  # ------------------------------------------------ df contains 
  # ----------------------B, S, T (the three components of alpha
  # ----------------------- as well as  q, and (nominal) p_value
  df = do.call(rbind, lapply(1:nrow(alpha_uni), function(jjj){
    alpha = alpha_uni[jjj, ]
    AAA = AAA_fn(alpha$B, alpha$S, alpha$T, MMM)
    davies_answer = davies_fn(zzz, mu_z, V_z, AAA)
    cbind(alpha, davies_answer)
  }))
  head(df)
  # ------------------------- basic is a data.frame that contains 
  # ------------------------ the quadratic form q and the p-value 
  # ----------------------- of the three basic statistics B, S, T
  basic = rbind(df[df$B == 1, ],
                df[df$S == 1, ],
                df[df$T == 1, ])
  rownames(basic) =  c("B", "S", "T")
  basic
  # ----------------------------------------------------- BS
  p_value_BS_fn = function(alpha_B){
    AAA = AAA_fn(alpha_B, 1 - alpha_B, 0, MMM)
    davies_fn(zzz, mu_z, V_z, AAA)$p_value
  }
  ooo = optimize(p_value_BS_fn, interval = c(0, 1))
  alpha_B = ooo$minimum
  BS_0 = data.frame(B = alpha_B, S = 1 - alpha_B, T = 0, p_value = ooo$objective)
  possibles = rbind(BS_0, basic[c("B", "S"), names(basic) != "q"])
  BS = possibles[which.min(possibles$p_value), ]
  # ----------------------------------------------------- BT
  p_value_BT_fn = function(alpha_B){
    AAA = AAA_fn(alpha_B, 0, 1 - alpha_B, MMM)
    davies_fn(zzz, mu_z, V_z, AAA)$p_value
  }
  ooo = optimize(p_value_BT_fn, interval = c(0, 1))
  alpha_B = ooo$minimum
  BT_0 = c(B = alpha_B, S = 0, T = 1 - alpha_B, p_value = ooo$objective)
  possibles = rbind(BT_0, basic[c("B", "T"), names(basic) != "q"])
  BT =  possibles[which.min(possibles$p_value), ]
  #------------------------------------------------------ ST
  p_value_ST_fn = function(alpha_S){
    AAA = AAA_fn(0, alpha_S, 1 - alpha_S, MMM)
    davies_fn(zzz, mu_z, V_z, AAA)$p_value
  }
  ooo = optimize(p_value_ST_fn, interval = c(0, 1))
  alpha_S = ooo$min
  ST_0 = c(0, S = alpha_S, T = 1 - alpha_S, p_value = ooo$objective)
  possibles = rbind(ST_0, basic[c("S", "T"), names(basic) != "q"])
  ST =  possibles[which.min(possibles$p_value), ]
  # ----------------------------------------------------- BST
  p_value_fn = function(theta){
    alpha = alpha_lu_fn(theta)
    AAA = AAA_fn(alpha[1], alpha[2], alpha[3], MMM)
    davies_fn(zzz, mu_z, V_z, AAA)$p_value
  }
  lu = optim(
    theta, p_value_fn,
    method = "L-BFGS-B", lower = rep(0, 3), upper = rep(pi/2, 2))
  counts_lu = lu$counts
  theta = lu$par
  BST_0 = c(alpha_lu_fn(theta), p_value = lu$value)
  possibles = rbind(BST_0, basic[, names(basic) != "q"])
  BST =  possibles[which.min(possibles$p_value), ]
  #----------------------------------------------------- optimized
  optimized_0 = as.data.frame(rbind(BS, BT, ST, BST))
  q = unlist(lapply(1:nrow(optimized_0), function(kkk){
    AAA = AAA_fn(optimized_0[kkk, "B"], optimized_0[kkk, "S"], optimized_0[kkk, "T"], MMM)
    as.vector(t(zzz) %*% AAA %*% zzz)
  }))
  xxx = x = unlist(lapply(optimized_0$p_value, q_alice_fn))
  optimized = cbind(optimized_0, q, x)[, c("B", "S", "T", "q", "p_value", "x")]
  rownames(optimized) =  c("BS", "BT", "ST", "BST")
  optimized
  #------------------------------------------------------- output
  p_value_basic = basic$p_value
  names(p_value_basic) = rownames(basic)
  p_value_basic
  
  q_basic = basic$q
  names(q_basic) = rownames(basic)
  q_basic
  
  q_optimized = optimized$q
  names(q_optimized) = rownames(optimized)
  q_optimized
  
  names(xxx) = rownames(optimized)
  list(
    #AAA_basic = AAA_basic,
    basic = basic,
    optimized = optimized,
    p_value_basic = p_value_basic,
    q_basic = q_basic,
    q_optimized = q_optimized,
    theta = theta,
    xxx = xxx)
}
```

  answer = if (rankMatrix(g_target) < MMM) {
    list(message = "error--g_target not full rank")
  } else {
    bo = basic_and_optimized_lu_fn(alpha_uni, g_target, Phi, theta_init, WWW, y_1, y_2)
    p_value_basic = bo$p_value_basic
    p_value_basic = bo$p_value_basic
    q_basic = bo$q_basic
    q_optimized = bo$q_optimized
    qqq = c(q_basic, q_optimized)
    theta = bo$theta
    x_observed = bo$xxx
    x_observed
    #--------------------------------------------------- p_value_optimized
    ooo = p_value_optimized_fn(
      adaptive_conf_level, alpha_uni, calculate_optimized, MMM, 
      N_sim_reps_interval, N_sim_reps_limit, Phi, sampling_set,
      test_size, theta, WWW, x_observed, y_1, y_2
    )
    so_far_so_good = ooo$so_far_so_good
    N_sim_reps_required = ooo$N_sim_reps_required
    p_value_optimized = ooo$p_value_optimized
    #-------------------------------------------------------------- output
    if(so_far_so_good){
      p_value = c(p_value_basic, p_value_optimized)
      list(
        N_sim_reps_required = N_sim_reps_required,
        p_value = p_value,
        q = qqq,
        x = x_observed)
    } else {
      list(message = "could not obtain full-rank g_target_sim matrices")
    }
  }
  answer
}
```

