
---
title: "gatars version 0.2.9"
author: ''
output:
  html_document:
    css: ~/Documents/headings.css
    fig_caption: yes
    highlight: default
    number_sections: yes
    theme: readable
    toc: yes
  pdf_document:
    highlight: tango
    number_sections: yes
    pandoc_args: --variable=geometry:margin=0.75in
    toc: yes
fontsize: 12pt
---
<A NAME="top"> </A>
```{r global_options, include=FALSE}
require(knitr)
opts_chunk$set(eval = TRUE, echo = TRUE, fig.height = 5, fig.width = 8, tidy=FALSE)
```

"2017-11-02 13:21:28 PDT"

# Introduction {#intro}

For comments or questions, please contact Gail Gong at gailgongster@gmail.com

`gatars` is an acronym for "Genetic Association Tests for Arbitrarily Related Subjects"

`gatars` tests the association between a specified set of $M$ genetic markers, called target 
markers, and a binary or quantitative trait using subjects with any genealogical 
relationship. The goal is to test the null hypothesis of no association between trait and 
genotypes at the M target markers.  We condition on trait phenotypes and treat genotypes
as random.  Test statistics include three "basic" statistics--the squared burden statistic $Q_B$,
the sequence kernel association test (SKAT) statistic $Q_S$, and a new trait-based kernel
statistic $Q_T$--as well as an ensemble of four statistics $Q_{BS}$, $Q_{BT}$, $Q_{ST}$, $Q_{BST}$ that optimize
linear combinations of the three basic statistics.  All seven statistics are summarized in 
Table 1 of the manuscript.

The test statistics use obervations on $N$ subjects, who are sampled based on their
phenotypes $y_n$ for a binary or quantitative trait, $n = 1, \cdots, N$.  The calculations require for
each subject a vector $g_n = \big(g_{n1}, \cdots, g_{nM} \big)$ of coded genotypes at $M$ target markers, and a
user-specified null trait prediction $\mu_n$ (for example $\mu_n$ can be the mean of the $N$ $y_n$ values
or the value predicted from a logistic or linear regression of $y_n$ on nongenetic covariates
and/or on principal components of ancestry.)  Define the genotype matrix to be the $N \times M$
matrix $G$ whose $(n,m)$-th component is the genotype $g_{nm}$ of the $n$-th subject at the $m$-th 
target marker.

The test statistics account for covariances among the components of $G$.  For pairwise 
intermarker covariances (summarized by the $M \times M$ matrix $\Gamma$ in the manuscript). `gatars` uses the
empirical covariances based on the genotypes of the $N$ subjects.  The user must provide the
pairwise inter-personal correlation coefficients (summarized by the $N \times N$ matrix $\Psi$ in the
manuscript).

Testing the null hypothesis for the optimized statistics requires resampling from sets of
markers located throughout the autosomal genome. `gatars` needs $M$ sampling sets, one for 
each target marker, with the sampling set for the $m$-th target marker containing markers
all of whose minor allele frequencies (MAFs) match that of target marker $m$.  The markers in
the sampling sets should be independent of the target markers and of any markers known
to be associated with the trait.  `gatars` can create these sampling sets provided the user
supplies two data sets containing: (1) the $N$ subjects' genotypes at roughly 100,000
markers located thoughout the autosomal genome; (2) the bp positions of any autosomal 
markers known to be trait-assoicated.  `gatars` creates the sampling sets by dividing the
human autosome into independent segments using the recombination hotspots identified
by Myers et al (2005).  To do so, `gatars` has translated the base-pair positions of these 
recombination hotspots to Build hg38/GRCh38.

[TOP](#top)

# Installing `gatars` {#install-gatars}

If you have not yet installed R, download the latest version (at least 3.2.4) for your operating system at:

http://www.r-project.org

Run the R application. To install the `gatars` package, enter the following lines of code to the R prompt:

```
install.packages("devtools")
library(devtools)
install_github("gailg/gatars")
if("gatars" %in% rownames(installed.packages())){
  print("gatars installed successfully--you are good to go!")
} else {
  print("something went wrong--ask for help")
}
```
If your installation was successful, you should see the message

```
[1] "gatars installed successfully--you are good to go!"
```

[TOP](#top)

# Data required {#data-required}

`gatars` requires quite large data, and typically, which can be organized in Plink (http://pngu.mgh.harvard.edu/~purcell/plink/).  We assume that you can bring your data into R and create the following data sets.

## `bim`

A `data.frame` containing the three columns `chromosome`, `snp`, `bp` and 
containing $L$ rows corresponding to the $M$ target markers and those used to build the
sampling sets.  The $l$-th row of `bim` summarizes the $l$-th marker and corresponds to the $l$-th column of `genotype`(described below). The column labeled `chromosome` contains integers between $1$ and $22$ (other integers may be included, but only chromosomes $1$ through $22$` will be used), the `snp` column contains character strings that name the marker, and the `bp` column is contains integers giving the markers' positions (bp).  (Because the `gatars` data set `hotspot` (described below) is in Build hg38/GRCh38, `bp` must also be expressed in Build hg38/GRCh38.) 

## `genotype`

An $N \times L$ matrix, whose $(n,l)$-th element records the $n$-th subject's coded genotype at the $l$-th
of $L$ markers. The $l$-th column of `genotype` corresponds to the $l$-th row of `bim`.  The object
`genotype` could be obtained by reading in the `.bed` file from plink.  (Here we distinguish
`genotype` (the $N \times L$ matrix whose columns correspond to the $M$ target markers AND the
additional $L - M$ potential sampling set markers) from the $N \times M$ matrix $G$, described in the manuscript, containing just the $M$ target 
markers.)

## `target_markers`

A vector that is a subset of the column `bim$snp`.  This vector identifies the target markers.

## `phen`

A `data.frame` containing $N$ rows and the two column `y` and `mu` where `y` contains the subjects' coded phenotypes (either `0` or `1` if the trait is binary or a real number if the trait is quantitative) and whre `mu` contains the user-specified null trait prediction (a real number) for $y$.

## `Psi` = $\Psi$

`Psi` is the $N \times N$ matrix of user-specified null correlation coefficients between pairs of subjects
at any marker.  `Psi` can be estimated using known family pedigree structures and/or from the
subjects' genotypes at markers independent of those in the target set in software packages
such as KING (Manichaikul et al 2010).


## `exclusion_region`

A `data.frame` with one or several rows of the three columns `chromosome`, `start`, and `end`. 
Each row of this `exclusion_region` reflects one contiguous genomic region known to 
be trait-associated and therefore a region used by `gatars` when creating the sampling sets.
The column `chromosome` is an integer between `1` and `22` identifying the autosomal
chromosome containing the region, and `start` and `end` describe its starting and ending 
positions (bp).  If the region consists of a single marker, then `start` and `end` are both equal 
to the position of this marker.  `start` and `end` must be expressed in Build hg38/GRCh38.

In addition, the `gatars` package provides the data set `hotspot` for your convenience.
`hotspot contains the columns `chromosome`, `start`, and `end`. `chromosome` gives the 
chromosome (an integer between $1$ and $22$); `start` and `end` give the extent of the recombination
hotspot (integers reflecting their Build hg38/GRCh38 bp positions).

# All the code in one place

```{r eval = TRUE}
library(gatars)
# Preparing the data
bim = alternative_example$bim
exclusion_region = alternative_example$exclusion_region
fam = alternative_example$fam
genotype = alternative_example$genotype
target_markers = alternative_example$target_markers[3:5]
Psi = alternative_example$Psi
NNN = nrow(fam)
first_ten = 1:10
last_ten = NNN - (9:0)
matrix_image_fn(Psi[c(first_ten, last_ten), c(first_ten, last_ten)],
                main = "First and last 10 rows and columns of Psi")

# Checking the rank of the genotype_target_markers matrix
library(Matrix)
genotype_target_markers = genotype[, target_markers]
list(target_markers = target_markers,
     rank = as.numeric(rankMatrix( genotype_target_markers)))

# Creating the sampling_set
set.seed(2)
epsilon_on_log_scale = 0.01
exclusion_region = NULL
sampling_set = gatars_sampling_set(
  bim,
  epsilon_on_log_scale,
  exclusion_region,
  genotype,
  hotspot,
  target_markers
)
print(sampling_set)

previous_sampling_set = sampling_set
set.seed(2)
exclusion_region = alternative_example$exclusion_region
head(exclusion_region)
sampling_set = gatars_sampling_set(
  bim,
  epsilon_on_log_scale,
  exclusion_region,
  genotype,
  hotspot,
  target_markers
)
print(previous_sampling_set)
print(sampling_set)

# Calling gatars 
start = Sys.time()
set.seed(1)
gatars(fam, Psi, sampling_set, N_sim_reps = 100, weights = c(5, 3, 2))
elapsed_time = Sys.time() - start
paste("100 sim reps used", 
      round(elapsed_time, 1), attributes(elapsed_time)$units)

start = Sys.time()
set.seed(1)
gatars(fam, Psi, sampling_set, N_sim_reps = 1000, weights = c(5, 3, 2))
elapsed_time = Sys.time() - start
paste("1000 sim reps used", 
      round(elapsed_time, 1), attributes(elapsed_time)$units)

start = Sys.time()
set.seed(1)
gatars(fam, Psi, sampling_set, N_sim_reps = 2000, weights = c(5, 3, 2))
elapsed_time = Sys.time() - start
paste("2000 sim reps used", 
      round(elapsed_time, 1), attributes(elapsed_time)$units)












































