
---
title: "gatars version 0.2.10"
author: ''
output:
  html_document:
    css: ~/Documents/headings.css
    fig_caption: yes
    highlight: default
    number_sections: yes
    theme: readable
    toc: yes
  pdf_document:
    highlight: tango
    number_sections: yes
    pandoc_args: --variable=geometry:margin=0.75in
    toc: yes
fontsize: 12pt
---
<A NAME="top"> </A>
```{r global_options, include=FALSE}
require(knitr)
opts_chunk$set(eval = TRUE, echo = TRUE, fig.height = 5, fig.width = 8, tidy=FALSE)
```

"2017-11-08 07:16:32 PST"

# Introduction {#intro}

For comments or questions, please contact Gail Gong at gailgongster@gmail.com

`gatars` is an acronym for "Genetic Association Tests for Arbitrarily Related Subjects"

`gatars` tests the association between a specified set of $M$ genetic markers, called target 
markers, and a binary or quantitative trait using subjects with any genealogical 
relationship. The goal is to test the null hypothesis of no association between trait and 
genotypes at the M target markers.  We condition on trait phenotypes and treat genotypes
as random.  Test statistics include three "basic" statistics--the squared burden statistic $Q_B$,
the sequence kernel association test (SKAT) statistic $Q_S$, and a new trait-based kernel
statistic $Q_T$--as well as an ensemble of four statistics $Q_{BS}$, $Q_{BT}$, $Q_{ST}$, $Q_{BST}$ that optimize
linear combinations of the three basic statistics.  All seven statistics are summarized in 
Table 1 of the manuscript.

The test statistics use obervations on $N$ subjects, who are sampled based on their
phenotypes $y_n$ for a binary or quantitative trait, $n = 1, \cdots, N$.  The calculations require for
each subject a vector $g_n = \big(g_{n1}, \cdots, g_{nM} \big)$ of coded genotypes at $M$ target markers, and a
user-specified null trait prediction $\mu_n$ (for example $\mu_n$ can be the mean of the $N$ $y_n$ values
or the value predicted from a logistic or linear regression of $y_n$ on nongenetic covariates
and/or on principal components of ancestry.)  Define the genotype matrix to be the $N \times M$
matrix $G$ whose $(n,m)$-th component is the genotype $g_{nm}$ of the $n$-th subject at the $m$-th 
target marker.

The test statistics account for covariances among the components of $G$.  For pairwise 
intermarker covariances (summarized by the $M \times M$ matrix $\Gamma$ in the manuscript). `gatars` uses the
empirical covariances based on the genotypes of the $N$ subjects.  The user must provide the
pairwise inter-personal correlation coefficients (summarized by the $N \times N$ matrix $\Psi$ in the
manuscript).

Testing the null hypothesis for the optimized statistics requires resampling from sets of
markers located throughout the autosomal genome. `gatars` needs $M$ sampling sets, one for 
each target marker, with the sampling set for the $m$-th target marker containing markers
all of whose minor allele frequencies (MAFs) match that of target marker $m$.  The markers in
the sampling sets should be independent of the target markers and of any markers known
to be associated with the trait.  `gatars` can create these sampling sets provided the user
supplies two data sets containing: (1) the $N$ subjects' genotypes at roughly 100,000
markers located thoughout the autosomal genome; (2) the bp positions of any autosomal 
markers known to be trait-assoicated.  `gatars` creates the sampling sets by dividing the
human autosome into independent segments using the recombination hotspots identified
by Myers et al (2005).  To do so, `gatars` has translated the base-pair positions of these 
recombination hotspots to Build hg38/GRCh38.

[TOP](#top)

# Installing `gatars` {#install-gatars}

If you have not yet installed R, download the latest version (at least 3.2.4) for your operating system at:

http://www.r-project.org

Run the R application. To install the `gatars` package, enter the following lines of code to the R prompt:

```
install.packages("devtools")
library(devtools)
install_github("gailg/gatars")
if("gatars" %in% rownames(installed.packages())){
  print("gatars installed successfully--you are good to go!")
} else {
  print("something went wrong--ask for help")
}
```
If your installation was successful, you should see the message

```
[1] "gatars installed successfully--you are good to go!"
```

[TOP](#top)

# Data required {#data-required}

`gatars` requires quite large data, and typically, which can be organized in Plink (http://pngu.mgh.harvard.edu/~purcell/plink/).  We assume that you can bring your data into R and create the following data sets.

## `bim`

A `data.frame` containing the three columns `chromosome`, `snp`, `bp` and 
containing $L$ rows corresponding to the $M$ target markers and those used to build the
sampling sets.  The $l$-th row of `bim` summarizes the $l$-th marker and corresponds to the $l$-th column of `genotype`(described below). The column labeled `chromosome` contains integers between $1$ and $22$ (other integers may be included, but only chromosomes $1$ through $22$` will be used), the `snp` column contains character strings that name the marker, and the `bp` column is contains integers giving the markers' positions (bp).  (Because the `gatars` data set `hotspot` (described below) is in Build hg38/GRCh38, `bp` must also be expressed in Build hg38/GRCh38.) 

## `genotype`

An $N \times L$ matrix, whose $(n,l)$-th element records the $n$-th subject's coded genotype at the $l$-th
of $L$ markers. The $l$-th column of `genotype` corresponds to the $l$-th row of `bim`.  The object
`genotype` could be obtained by reading in the `.bed` file from plink.  (Here we distinguish
`genotype` (the $N \times L$ matrix whose columns correspond to the $M$ target markers AND the
additional $L - M$ potential sampling set markers) from the $N \times M$ matrix $G$, described in the manuscript, containing just the $M$ target 
markers.)

## `target_markers`

A vector that is a subset of the column `bim$snp`.  This vector identifies the target markers.

## `phen`

A `data.frame` containing $N$ rows and the two column `y` and `mu` where `y` contains the subjects' coded phenotypes (either `0` or `1` if the trait is binary or a real number if the trait is quantitative) and whre `mu` contains the user-specified null trait prediction (a real number) for $y$.

## `Psi` = $\Psi$

`Psi` is the $N \times N$ matrix of user-specified null correlation coefficients between pairs of subjects
at any marker.  `Psi` can be estimated using known family pedigree structures and/or from the
subjects' genotypes at markers independent of those in the target set in software packages
such as KING (Manichaikul et al 2010).


## `exclusion_region`

A `data.frame` with one or several rows of the three columns `chromosome`, `start`, and `end`. 
Each row of this `exclusion_region` reflects one contiguous genomic region known to 
be trait-associated and therefore a region used by `gatars` when creating the sampling sets.
The column `chromosome` is an integer between `1` and `22` identifying the autosomal
chromosome containing the region, and `start` and `end` describe its starting and ending 
positions (bp).  If the region consists of a single marker, then `start` and `end` are both equal 
to the position of this marker.  `start` and `end` must be expressed in Build hg38/GRCh38.

In addition, the `gatars` package provides the data set `hotspot` for your convenience.
`hotspot contains the columns `chromosome`, `start`, and `end`. `chromosome` gives the 
chromosome (an integer between $1$ and $22$); `start` and `end` give the extent of the recombination
hotspot (integers reflecting their Build hg38/GRCh38 bp positions).

[TOP](#top)

# An example data set {#example-data-set}

For illustration I will now use the example data set provided by `gatars`.
You can access this example data with the following commands.
```{r}
library(gatars)
bim = gatars_example$bim
exclusion_region = gatars_example$exclusion_region
genotype = gatars_example$genotype
phenotype = gatars_example$phenotype
Psi = gatars_example$Psi
target_markers = gatars_example$target_markers[3:5]
```

## `bim` has `r nrow(bim)` rows and includes the columns `chromosome`, `snp`, and  `bp`

```{r}
str(bim)
```

## `genotype` has `r nrow(genotype)` rows and `r ncol(genotype)` columns

I chose a relatively small number of columns to keep the example small.

```{r}
str(genotype)
```

The column names of `genotype` must match `bim$snp`:

```{r}
all(colnames(genotype) == bim$snp)
```

## `target_markers` is a character vector containing `r length(target_markers)` names

```{r}
str(target_markers)
```

The elements in `target_markers` must be included in `bim$snp`

```{r}
all(target_markers %in% bim$snp)
```

## `phenotype` has `r nrow(phenotype)` rows and includes the columns `y` and `mu`

```{r}
str(phenotype)
```

## `Psi` is a `r nrow(Psi)` $\times$ `r nrow(Psi)`  square matrix

```{r}
str(Psi)
```

The following figure reflects the fact that the first `100` people are made up of `50` sib pairs, and the last `100` people` are independent.

```{r Psi_picture, fig.height = 4, fig.width = 5}
NNN = nrow(phenotype)
first_ten = 1:10
last_ten = NNN - (9:0)
matrix_image_fn(Psi[c(first_ten, last_ten), c(first_ten, last_ten)],
                main = "First and last 10 rows and columns of Psi")
```


## `exclusion_region` is a `data.frame` containing  the columns `chromosome`, `begin`, and `end`

```{r}
str(exclusion_region)
```

## `hotspot` is available as soon as you enter `gatars`:

```{r}
str(hotspot)
```


# Check that your genotype matrix has full rank

`genotype_target_markers` is the $N \times M$ genotype matrix.

```{r}
library(Matrix)
genotype_target_markers = genotype[, target_markers]
list(target_markers = target_markers,
     rank = as.numeric(rankMatrix( genotype_target_markers)))
```

# gatars-sampling-set {#gatars-sampling-set}

Once you have your data in R, and you have checked that the column rank of your genotype matrix $G =$ `genotype_target_markers` is $M$, use the function `gatars_sampling_set`  to create your sampling sets. 

## A description of how `gatars_sampling_set` builds the sampling sets

Each column in `genotype` corresponds to a row in `bim`, and both correspond to a marker. After removing the target markers, we want to use the remaining markers to form $M$ sampling sets, one for each of the target markers. The two requirements for the markers in a sampling set are: 1) that they have MAFs matching that of the target marker; and 2) that they are independent of all target markers and all markers known to be trait-associated.

We assume that the recombination hotspots divide the autosomal genome into independent segments, and so markers within two consecutive hotspots are independent of those within another two consecutive hotspots.  Also, we remove all segments containing any target markers or any markers defined by the `exclusion_region` data set.

We calculate the empirical MAFs of the markers in the remaining segments, say that a marker's MAF matches the MAF $\pi_m$ of the $m$-th target marker if it is within
$\pi_m \times \big[1 -$ `epsilon` $, 1 +$ `epsilon` $\big]$.  If the number of markers satisfying the matching requirement exceeds `1000`, `gatars` randomly chooses `1000`.


## The arguments of `gatars_sampling_set`

The following is the function header of `gatars_sampling_set` showing which arguments are needed and in which order.

```
gatars_sampling_set(
  bim,
  epsilon,
  exclusion_region,
  genotype,
  hotspot,
  target_markers
)
```
### `bim`,  `exclusion_region`, `genotype`, `hotspot`, and `target_markers`,

These objects have already been discussed in [An example data set](#example-data-set).

### `epsilon` 

A positive small real number used to parametrize the matching.

When creating the $m$-th sampling set for target marker with minor allele frequency $\pi_m$, only those markers whose minor allele frequencies falling within the interval
$\pi_m \times \big[1 -$ `epsilon` $, 1 +$ `epsilon` $\big]$ 
can be included in the sampling set.

## Example calls to `gatars_sampling_set` 

```{r eval = TRUE}
set.seed(42)
epsilon = 0.01
exclusion_region = NULL
sampling_set = gatars_sampling_set(
  bim,
  epsilon,
  exclusion_region,
  genotype,
  hotspot,
  target_markers
)
print(sampling_set)
```

```{r eval = TRUE}
previous_sampling_set = sampling_set
set.seed(42)
exclusion_region = gatars_example$exclusion_region
head(exclusion_region)
sampling_set = gatars_sampling_set(
  bim,
  epsilon,
  exclusion_region,
  genotype,
  hotspot,
  target_markers
)
print(previous_sampling_set)
print(sampling_set)
```

[TOP](#top)

# `gatars`

## A description of `gatars`

`gatars` calculates the p-values of the following seven statistics: the squared burden $Q_B$, the SKAT $Q_S$, the case based $Q_T$, and optimal linear combinations $Q_{BS}$, $Q_{BT}$, $Q_{ST}$, $Q_{BST}$. 
The p-value of any linear combination of the basic statistics $Q_B$, $Q_S$, and $Q_T$  can be calculated using the `davies` function from the `CompQuadForm` package, but this theory no longer applies when a statistic is an optimal one in a universe of linear combinations. We obtain the null distribution of these optimal statistics by resorting to an innovative kind of simulation, which we call _Genome Resampling_. 

Recall our notation:  $G$ is the $N \times M$ genotype matrix of target markers,  $y$ is the $N$-vector indicator for disease or a quantative measurement, and  `e_y` is an $N$-vector of trait predictions.
Any linear combination of the basic statistics can be written as a quadratic form $Q(\alpha) = Z^T A(\alpha) Z$ where $Z$ is a linear function of the two vectors $W G y$ and $W G p$, $W$ is a diagonal matrix whose nonzero elements weight the target markers, $A(\alpha)$ is a square matrix that does not depend on $G$, and $\alpha$ are the coeffients of the linear combination of basic statistics.  $Z$ is a multivariate normal vector when $N$ is large, with mean $\mu$ and covariance matrix $\Sigma$ which depend on the first two moments of $G$. For a fixed value of $\alpha$, the `davies` function can calculate the p-value for $Q(\alpha)$. Each optimal statistic $Q_{BS}$, $Q_{BT}$, $Q_{ST}$, $Q_{BST}$ is a monotonoic (decreasing) function, $-\text{log}_{10}$, of the smallest p-value of $Q(\alpha)$ over a set of values of $\alpha$. 

To run a simulation to get the null distribution of an optimal statistic $Q_{\text{optimal}}$, we need to obtain simulated observations of a large number $S$ of replications of the genotype matrix $\Big\{\tilde{G}^{(1)}, \cdots, \tilde{G}^{(S)}\Big\}$. On the $s$-th simulated genotype matrix $\tilde{G}_s$, we calculate $Q(\alpha)$ over the set of values of $\alpha$ and obtain the optimal one $Q^{(s)}_{\text{optimal}}$. The p-value of $Q_{\text{optimal}}$ is the proportion of $\Big\{ Q^{(1)}_{\text{optimal}}, \cdots, Q^{(S)}_{\text{optimal}} \Big\}$ that exceed $Q_{\text{optimal}}$.

The question becomes how to generate each $\tilde{G}^{(s)}$ so that it has the same distribution as the observed genotype matrix $G$ conditional on its following the null model? We propose _Genome Resampling_: Form $\tilde{G}^{(s)}$ by sampling one column from each sampling set created according to Section 
[gatars-sampling-set](#gatars-sampling-set). 

## The arguments of `gatars`

The following is the function header of `gatars` showing which arguments are needed and in which order.

```
gatars(phenotype, Psi, sampling_set, N_sim_reps, weights)
```

### `phenotype` and `Psi` 

These objects have already been discussed in [An example data set](#example-data-set).

### `sampling_set` 

This object is the result of `gatars_sampling_set` and you may follow the example from [gatars-sampling-set](#gatars-sampling-set)

### `N_sim_reps

An positive integer which specifies the number of replications in the something.


### `weights`

A vector of length `MMM = 3`.  The entries in the vector are non-negative real numbers.  The size of the `m`-th entry reflects the importance of the `m`-th target marker. If `weights` is not specified, the function assume you would like equal weights among the `MMM` target markers.

[TOP](#top)

## Example calls to `gatars`




# All the code in one place

```{r eval = TRUE}
library(gatars)
# Preparing the data
bim = gatars_example$bim
exclusion_region = gatars_example$exclusion_region
genotype = gatars_example$genotype
phenotype = gatars_example$phenotype
Psi = gatars_example$Psi
target_markers = gatars_example$target_markers[3:5]
NNN = nrow(phenotype)
first_ten = 1:10
last_ten = NNN - (9:0)
matrix_image_fn(Psi[c(first_ten, last_ten), c(first_ten, last_ten)],
                main = "First and last 10 rows and columns of Psi")

# Checking the rank of the genotype_target_markers matrix
library(Matrix)
genotype_target_markers = genotype[, target_markers]
list(target_markers = target_markers,
     rank = as.numeric(rankMatrix( genotype_target_markers)))

# Creating the sampling_set
set.seed(42)
epsilon = 0.01
exclusion_region = NULL
sampling_set = gatars_sampling_set(
  bim,
  epsilon,
  exclusion_region,
  genotype,
  hotspot,
  target_markers
)
print(sampling_set)

previous_sampling_set = sampling_set
set.seed(42)
exclusion_region = gatars_example$exclusion_region
head(exclusion_region)
sampling_set = gatars_sampling_set(
  bim,
  epsilon,
  exclusion_region,
  genotype,
  hotspot,
  target_markers
)
print(previous_sampling_set)
print(sampling_set)

# Calling gatars 
start = Sys.time()
set.seed(42)
gatars(phenotype, Psi, sampling_set, N_sim_reps = 100, weights = c(5, 3, 2))
elapsed_time = Sys.time() - start
paste("100 sim reps used", 
      round(elapsed_time, 1), attributes(elapsed_time)$units)

start = Sys.time()
set.seed(42)
gatars(phenotype, Psi, sampling_set, N_sim_reps = 1000, weights = c(5, 3, 2))
elapsed_time = Sys.time() - start
paste("1000 sim reps used", 
      round(elapsed_time, 1), attributes(elapsed_time)$units)

start = Sys.time()
set.seed(42)
gatars(phenotype, Psi, sampling_set, N_sim_reps = 2000, weights = c(5, 3, 2))
elapsed_time = Sys.time() - start
paste("2000 sim reps used", 
      round(elapsed_time, 1), attributes(elapsed_time)$units)

```










































