---
title: "04-sampling_set_fn"
author: ""
output:
  html_document:
    css: ~/Documents/headings.css
    fig_caption: yes
    highlight: default
    number_sections: yes
    theme: readable
    toc: yes
  pdf_document:
    highlight: tango
    number_sections: yes
    pandoc_args: --variable=geometry:margin=0.75in
    toc: no
fontsize: 10pt
---
<A NAME="top"> </A>

```{r global_options, include=FALSE}
require(knitr)
opts_chunk$set(eval = TRUE, echo = TRUE, fig.height = 10, fig.width = 8, tidy=FALSE)
```

`r getwd()`  

"2017-09-14 08:12:36 PDT"

Last knit was `r Sys.time()`

```{r}
library(gatars)
bim = alternative_example$bim
epsilon_on_log_scale = 0.02
exclusion_region = data.frame(chromosome = integer(), begin = integer(), end = integer())
genotype = alternative_example$genotype
target_markers = alternative_example$target_markers[3:5]
params_sampling_set = params_sampling_set_fn(
  bim, epsilon_on_log_scale, exclusion_region,
  genotype, hotspot, target_markers)
names(params_sampling_set)
```


```{r}
bim_with_target_regions_removed = bim_with_target_regions_removed_fn(params_sampling_set)
str(bim_with_target_regions_removed)
```

```{r}
#' @export
bim_with_target_regions_removed_fn = function(params_sampling_set){
  bim = params_sampling_set$bim
  hotspot = params_sampling_set$hotspot
  target_markers = params_sampling_set$target_markers
  target_bim = bim[bim$snp %in% target_markers, ]
  target_part = with(target_bim, data.frame(chromosome, begin = bp, end = bp))
  exclusion_region = rbind(params_sampling_set$exclusion_region, target_part)
  exclusion_chromosome = unique(exclusion_region$chromosome)
  chromosomes_with_target_and_exlusion_regions_removed = lapply(1:22, function(chromosome){  
    answer = if(chromosome %in% exclusion_chromosome){
      segments_0 = independent_segment_0_fn(chromosome, bim, hotspot)
      eee = exclusion_region[exclusion_region$chromosome == chromosome, ]
      intersect_with_genotype_q = sapply(segments_0, function(this){
        any(unlist(lapply(1:nrow(eee), function(kkk){
          any(eee[kkk, ]$begin : eee[kkk, ]$end %in% this$bp)
        })))
      })
      do.call(rbind, segments_0[!intersect_with_genotype_q])
    } else {
      bim[bim$chromosome == chromosome, ]
    }
    answer
  })
  do.call(rbind, chromosomes_with_target_and_exlusion_regions_removed)
}

# "2016-10-19 12:35:20 PDT" handles exclusion_region
```

