---
title: "05-bim_with_target_exclusion_regions_removed_fn"
author: ""
output:
  html_document:
    css: ~/Documents/headings.css
    fig_caption: yes
    highlight: default
    number_sections: yes
    theme: readable
    toc: yes
  pdf_document:
    highlight: tango
    number_sections: yes
    pandoc_args: --variable=geometry:margin=0.75in
    toc: no
fontsize: 10pt
---
<A NAME="top"> </A>

```{r global_options, include=FALSE}
require(knitr)
opts_chunk$set(eval = TRUE, echo = TRUE, fig.height = 10, fig.width = 8, tidy=FALSE)
```

`r getwd()`  

"2017-09-16 07:35:24 PDT"

Last knit was `r Sys.time()`

# `exclusion_region = NULL`

```{r}
library(gatars)
#--------------- Get ready to call the function
bim = alternative_example$bim
epsilon_on_log_scale = 0.02
table(bim$chromosome)
bim_for_chromosome_21 = bim[bim$chromosome == 21, ]
head(bim_for_chromosome_21, 20)
target_markers = c("exm-rs1006899",  "exm1564039")
bim[bim$snp %in% target_markers, ]
exclusion_region = NULL
genotype = alternative_example$genotype
params_sampling_set = params_sampling_set_fn(
  bim, epsilon_on_log_scale, exclusion_region,
  genotype, hotspot, target_markers)
names(params_sampling_set)
#--------------- Call the function
bim_with_target_and_exclusion_regions_removed = 
  bim_with_target_and_exclusion_regions_removed_fn(params_sampling_set)
#--------------- Notice that some lines have been removed
str(bim_with_target_and_exclusion_regions_removed)
str(bim)
#--------------- Focus on chromosome 21 because it contains the fewest snps
before = bim_for_chromosome_21 = bim[bim$chromosome == 21, ]
str(before)
after = bim_with_target_and_exclusion_regions_removed_for_chromosome_21 = 
  bim_with_target_and_exclusion_regions_removed[
    bim_with_target_and_exclusion_regions_removed$chromosome == 21, ]
str(after)
#--------------- These are the bp that have been removed
setdiff(before$bp, after$bp)
#--------------- Does this make sense?
hhh = hotspot_for_chromosome_21 = hotspot[hotspot$chromosome == 21, ]
bim[bim$snp %in% target_markers, ]
hhh[15600000 < hhh$center & hhh$center < 16000000, ]
bim_for_target_markers = bim[bim$snp %in% target_markers, ]
to_the_left = hhh[hhh$end < bim_for_target_markers[1, ]$bp, ]
to_the_left[nrow(to_the_left), ]
to_the_right = hhh[bim_for_target_markers[1, ]$bp < hhh$start, ]
to_the_right[1, ]
bim[bim$snp %in% target_markers, ]
bim_remove_due_to_first_target_marker = bim_for_chromosome_21[
  to_the_left[nrow(to_the_left), ]$end <= bim_for_chromosome_21$bp &
  bim_for_chromosome_21$bp <= to_the_right[1, ]$start, ]
#--------------- The part of bim that are removed due to the 
#----------------first target marker
bim_remove_due_to_first_target_marker  
#--------------- Repeat the above for the second_target_marker.
#--------------- This shows that sometimes not just the target
#--------------- marker is removed.
to_the_left = hhh[hhh$end < bim_for_target_markers[2, ]$bp, ]
to_the_left[nrow(to_the_left), ]
to_the_right = hhh[bim_for_target_markers[2, ]$bp < hhh$start, ]
to_the_right[1, ]
bim_remove_due_to_second_target_marker = bim_for_chromosome_21[
  to_the_left[nrow(to_the_left), ]$end <= bim_for_chromosome_21$bp &
  bim_for_chromosome_21$bp <= to_the_right[1, ]$start, ]
bim_remove_due_to_second_target_marker  
#--------------- The function also removes the parts of bim
#--------------- that fall inside the hotspots
bim_removed_because_they_are_in_hotspots = bim_for_chromosome_21[
  sapply(bim_for_chromosome_21$bp, function(this){
  any(hhh$start <= this & this <= hhh$end)
}), ]
bim_removed_because_they_are_in_hotspots$bp
#--------------- Check to see if the parts removed because of the 
#--------------- target markers and those removed because they fall
#--------------- within hotspots match those removed in the 
#--------------- original function call
all.equal(sort(c(bim_remove_due_to_first_target_marker$bp,
                 bim_remove_due_to_second_target_marker$bp,
                 bim_removed_because_they_are_in_hotspots$bp)),
          setdiff(before$bp, after$bp))
```




